go/carmen/memory_footprint.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package carmen
+
+import "github.com/Fantom-foundation/Carmen/go/common"
+
+type MemoryFootprint interface {
+	// GetChild returns a child of the memory footprint with the given name.
+	GetChild(name string) MemoryFootprint
+
+	// Value provides the amount of bytes consumed by the database structure (excluding its subcomponents)
+	Value() uintptr
+
+	// Total provides the amount of bytes consumed by the database structure including all its subcomponents
+	Total() uintptr
+
+	// ToString provides the memory footprint as a tree summary in a string
+	// The name param allows to give a name to the root of the tree.
+	ToString(name string) string
+
+	// String allow memory footprints to be used in format strings.
+	String() string
+
+	// Visit iterates the footprint tree (including all subcomponents) and applies visit onto them
+	Visit(visit func(footprint MemoryFootprint))
```

Comment:
I think the only reason why the wrapper `memoryFootprint` is needed is this recursive dependency to `MemoryFootprint`. If the callback returns e.g. a tuple `depth, name, size` the wrapper could be avoided and the `carmen` package would define only the interface.

Also the parameter `depth` would be very useful, otherwise the user does not know where he is currently in the tree.

As I said in previous comments round, it would be very useful to create an example how to use this new API - it would quickly discover all shortcomings